### PR TITLE
[GOAL2-877] Fix data race in asyncVoteVerifier during shutdown

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary
Fix the data race. 
The cause for this data race is that we were closing the output channel where the execution pool would write the result before clearing out the execution pool queue.

## Solution
Account for each entry that goes into the execution pool, and wait until all the items are executed before closing the output channel.

## Stacktrace
`algod(24485) : WARNING: DATA RACE
algod(24485) : Write at 0x00c00024fe70 by 
algod(24485) : goroutine 129:
algod(24485) :   runtime.closechan()
algod(24485) :       /home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/chan.go:334 +0x0
algod(24485) :   github.com/algorand/go-algorand/agreement.(*AsyncVoteVerifier).Quit()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/agreement/asyncVoteVerifier.go:155 +0xbd
algod(24485) :   github.com/algorand/go-algorand/agreement.(*Service).demuxLoop()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/agreement/service.go:155 +0x280
algod(24485) : Previous read at 0x00c00024fe70 by 
algod(24485) : goroutine 62:
algod(24485) :   runtime.chansend()
algod(24485) :       /home/travis/.gimme/versions/go1.12.linux.amd64/src/runtime/chan.go:142 +0x0
algod(24485) :   github.com/algorand/go-algorand/util/execpool.(*pool).worker()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/util/execpool/pool.go:162 +0x1bf
algod(24485) : Goroutine 129 (running) created at:
algod(24485) :   github.com/algorand/go-algorand/agreement.(*Service).Start()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/agreement/service.go:127 +0x1a2
algod(24485) :   github.com/algorand/go-algorand/node.(*AlgorandFullNode).Start()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/node/node.go:309 +0x1ba
algod(24485) :   github.com/algorand/go-algorand/daemon/algod.(*Server).Start()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/daemon/algod/server.go:152 +0x1c5
algod(24485) :   main.main()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/cmd/algod/main.go:289 +0x175a
algod(24485) : Goroutine 62 (running) created at:
algod(24485) :   github.com/algorand/go-algorand/util/execpool.MakePool()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/util/execpool/pool.go:85 +0x250
algod(24485) :   github.com/algorand/go-algorand/node.MakeFull()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/node/node.go:188 +0xfee
algod(24485) :   github.com/algorand/go-algorand/daemon/algod.(*Server).Initialize()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/daemon/algod/server.go:121 +0xb12
algod(24485) :   main.main()
algod(24485) :       /home/travis/gopath/src/github.com/algorand/go-algorand/cmd/algod/main.go:268 +0x1436`

### CODE REVIEW

**Comment by Vervious** [comment]
Interesting: https://github.com/golang/go/issues/27769

File: agreement/asyncVoteVerifier.go [hunk 1]
```diff
@@ -76,19 +77,19 @@ func MakeAsyncVoteVerifier(verificationPool execpool.BacklogPool) *AsyncVoteVeri
...
 
-	verifier.wg.Add(1)
+	verifier.workerWaitCh = make(chan struct{})
 	go verifier.worker()
 	return verifier
 }
 
 func (avv *AsyncVoteVerifier) worker() {
-	defer avv.wg.Done()
+	defer func() { close(avv.workerWaitCh) }()
```
Comment (Karmastic) [inline agreement/asyncVoteVerifier.go]: Why do you need to wrap in func()?  Won't `defer close(avv.workerWaitCh)` do the same thing?

File: agreement/asyncVoteVerifier.go [hunk 2]
```diff
@@ -129,7 +130,10 @@ func (avv *AsyncVoteVerifier) verifyVote(verctx context.Context, l LedgerReader,
 	default:
 		// if we're done while waiting for room in the requests channel, don't queue the request
 		req := asyncVerifyVoteRequest{ctx: verctx, l: l, uv: &uv, index: index, message: message, out: out}
-		avv.backlogExecPool.EnqueueBacklog(avv.ctx, avv.executeVoteVerification, req, avv.execpoolOut)
+		avv.wg.Add(1)
+		if avv.backlogExecPool.EnqueueBacklog(avv.ctx, avv.executeVoteVerification, req, avv.execpoolOut) != nil {
+			avv.wg.Done()
```
Comment (Karmastic) [inline agreement/asyncVoteVerifier.go]: Can you add a comment here explaining why we're calling .Done() only when this returns nil?

File: agreement/asyncVoteVerifier.go (hunk 1 above)
Comment (tsachiherman) [reply agreement/asyncVoteVerifier.go]: yes; updated.

File: agreement/asyncVoteVerifier.go [hunk 3]
```diff
@@ -141,20 +145,28 @@ func (avv *AsyncVoteVerifier) verifyEqVote(verctx context.Context, l LedgerReade
 	default:
 		// if we're done while waiting for room in the requests channel, don't queue the request
 		req := asyncVerifyVoteRequest{ctx: verctx, l: l, uev: &uev, index: index, message: message, out: out}
-		avv.backlogExecPool.EnqueueBacklog(avv.ctx, avv.executeEqVoteVerification, req, avv.execpoolOut)
+		avv.wg.Add(1)
+		if avv.backlogExecPool.EnqueueBacklog(avv.ctx, avv.executeEqVoteVerification, req, avv.execpoolOut) != nil {
+			avv.wg.Done()
```
Comment (Karmastic) [inline agreement/asyncVoteVerifier.go]: Same here

File: agreement/asyncVoteVerifier.go (hunk 2 above)
Comment (tsachiherman) [reply agreement/asyncVoteVerifier.go]: sure.

File: agreement/asyncVoteVerifier.go (hunk 3 above)
Comment (tsachiherman) [reply agreement/asyncVoteVerifier.go]: np.

**Pull request merged**